### PR TITLE
CallPromotionUtils: Correctly use IndexSize when determining the bit width of pointer offsets.

### PR DESCRIPTION
**Pull request by resistor**
This reapplies #119138 with a defensive fix for the assertion failure when building libcxx.
Unfortunately the failure does not reproduce on my machine, so I am not able to extract a test case.

The key insight for the fix comes from Jessica Clarke, who observes that `VTablePtr` may, in fact,
not be a pointer on return from `FindAvailableLoadedValue`.

Co-authored-by: Alexander Richardson <alexander.richardson@cl.cam.ac.uk>

### CODE REVIEW

File: llvm/lib/Transforms/Utils/CallPromotionUtils.cpp [hunk 1]
```diff
@@ -712,7 +712,12 @@ bool llvm::tryPromoteCall(CallBase &CB) {
       VTablePtrLoad, VTablePtrLoad->getParent(), BBI, 0, nullptr, nullptr);
   if (!VTablePtr)
     return false; // No vtable found.
-  APInt VTableOffsetGVBase(DL.getTypeSizeInBits(VTablePtr->getType()), 0);
+
+  // VTablePtr may not actually have a pointer type, so we can't find the
+  // address space through it. However, we know that the address space must be
+  // the same one used to load from the vtable in VTablePtrLoad.
```
Comment (jrtc27) [inline llvm/lib/Transforms/Utils/CallPromotionUtils.cpp:718]: I wouldn't rely on that, e.g. https://reviews.llvm.org/D114533 wanted to relax that restriction in some cases. Also, VTablePtrLoad->getPointerAddressSpace() is the address space of the pointer to the vtable pointer, not the vtable pointer itself (which would be VTablePtrLoad->getType()->getPointerAddressSpace()). At least I think what I said is correct, I confused myself several times writing this.

That is, we're dealing with:

```
  %object_base = alloca ptr
  %object = %object_base

  %vtable_ptr = getelementptr i8, ptr @gv, iN %vtable_gv_offset
  store ptr %vtable_ptr, ptr %object

  %vtable_ptr_load = load ptr, ptr %object
  %vtable_base_ptr = %vtable_ptr_load
  %vtable_entry_ptr = getelementptr i8, ptr %vtable_base_ptr, iN %vtable_offset
  %vtable_entry_load = load ptr, ptr %vtable_entry_ptr
  %callee = %vtable_entry_load
  call ptr %callee
```

Where we want vtable_ptr's address space, not object's.

Comment (jrtc27) [reply llvm/lib/Transforms/Utils/CallPromotionUtils.cpp:718]: I'd just add || !VTablePtr->isPointerType() to line 713

Comment (resistor) [reply llvm/lib/Transforms/Utils/CallPromotionUtils.cpp:718]: Revised to implement this fix instead.

File: llvm/lib/Transforms/Utils/CallPromotionUtils.cpp [hunk 2]
```diff
@@ -710,9 +710,13 @@ bool llvm::tryPromoteCall(CallBase &CB) {
...
   Value *VTablePtr = FindAvailableLoadedValue(
       VTablePtrLoad, VTablePtrLoad->getParent(), BBI, 0, nullptr, nullptr);
-  if (!VTablePtr)
+  if (!VTablePtr || !VTablePtr->getType()->isPointerTy())
     return false; // No vtable found.
-  APInt VTableOffsetGVBase(DL.getTypeSizeInBits(VTablePtr->getType()), 0);
+
+  // VTablePtr may not actually have a pointer type, so we can't find the
+  // address space through it. However, we know that the address space must be
+  // the same one used to load from the vtable in VTablePtrLoad.
```
Comment (jrtc27) [inline llvm/lib/Transforms/Utils/CallPromotionUtils.cpp:718]:
```suggestion
```
Comment is obsolete

**Comment by jrtc27** [comment]
> jrtc

jrtc27 (or my name)

File: llvm/lib/Transforms/Utils/CallPromotionUtils.cpp (hunk 2 above)
Comment (resistor) [reply llvm/lib/Transforms/Utils/CallPromotionUtils.cpp]: Done.

**Comment by resistor** [comment]
> > jrtc
> 
> jrtc27 (or my name)

Done, though I had to squash and rewrite history, which may make the PR reviews a bit hard to follow.

**Comment by jrtc27** [comment]
> > > jrtc
> > 
> > 
> > jrtc27 (or my name)
> 
> Done, though I had to squash and rewrite history, which may make the PR reviews a bit hard to follow.

LLVM's repo has 'Squash and merge" (the only option enabled) configured to pick the final commit message from the PR's title and description, so in future you don't actually need to go and rewrite the message(s) of the underlying commit(s).

**Review by jrtc27 — APPROVED** [review]
This looks good to me. If the buildbot is unhappy again with the modified version we should make sure to actually reproduce the issue (I can have a go another day if needed).
